nixos/modules/security/gnupg.nix

```diff
@@ -0,0 +1,253 @@
+            is provided in the shell variable <literal>secret_file</literal>.
+          '';
+        };
+      };
+    }));
+  };
+  enableGpgAgent = lib.mkEnableOption ''gpg-agent for decrypting secrets.
+
+    Otherwise, you'll have to forward an <literal>agent-extra-socket</literal>:
+    <screen>
+    <prompt>$ </prompt>ssh -nNT root@example.org -o StreamLocalBindUnlink=yes -R /var/lib/gnupg/S.gpg-agent:$(gpgconf --list-dirs agent-extra-socket)
+    </screen>
+  '' // {default = true; example = false;};
+  gpgAgentFlags = lib.mkOption {
+    type = types.listOf types.str;
+    default = [
+      "--default-cache-ttl" "600"
+      "--max-cache-ttl" "7200"
+    ];
+    description = ''
+      Extra flags passed to the <literal>gpg-agent</literal>
+      used to decrypt secrets.
+    '';
+  };
+};
+config = lib.mkIf (cfg.secrets != {}) {
+  systemd.sockets."gpg-agent" = lib.optionalAttrs cfg.enableGpgAgent {
+    description = "Socket for gpg-agent";
+    wantedBy = ["sockets.target"];
+    socketConfig.ListenStream = "${gnupgHome}/S.gpg-agent";
+    socketConfig.SocketMode = "0600";
+  };
+  environment.systemPackages =
+    # TODO: maybe this would be better to do that directly in pkgs.gnupg?
+    let gpgPresetPassphrase = pkgs.runCommand "gpg-preset-passphrase"
+      { preferLocalBuild = true;
+        allowSubstitutes = false;
+      } ''
+      mkdir -p $out/bin
+      ln -s -t $out/bin ${pkgs.gnupg}/libexec/gpg-preset-passphrase
+    ''; in
+    [ pkgs.gnupg gpgPresetPassphrase ];
+  systemd.services =
+    lib.optionalAttrs cfg.enableGpgAgent {
+      gpg-agent = {
+        description = "gpg-agent for decrypting GnuPG-protected secrets";
+        requires = ["gpg-agent.socket"];
+        serviceConfig = {
+          Type = "simple";
+          # Because /run/user/0 is wiped out by pam_systemd when root logouts,
+          # systemd.services.gpg-agent cannot put its socket in
+          # the path expected by gpg:
+          # /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
+          # derived from ${gnupgHome}.
+          #
+          # But GPG_AGENT_INFO is ignored with gpg >= 2.1,
+          # hence this hack makes gpg connect to ${gnupgHome}/S.gpg-agent
+          # by relaxing permissions of the expected socket directory.
+          # With this hack, uploading the passphrase should now be doable with just:
+          # ssh root@example.org 'gpg-preset-passphrase --homedir /var/lib/gnupg --preset $keygrip'
+          #
+          # An other way would be to remove /run/user/0 with:
+          # BindPaths = ["/var/empty:/run/user"];
+          # but this would also require a more complex passphrase uploading:
+          # ssh root@example.org "unshare --mount sh -c '
+          #   mount --bind /var/empty /run/user;
+          #   gpg-preset-passphrase --homedir /var/lib/gnupg --preset $keygrip'"
+          ExecStartPre = "${pkgs.coreutils}/bin/install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts";
```

Comment:
I think there might be a problem here: `/run/user/0` is a `tmpfs`. So its not just about deleting files but the whole filesystem is remounted. As a result this might lead to race conditions where a root user logins or out during deployment where the fs is mounted again.
If you really want to depend on an existing user-configured gpg-daemon (which I would not recommend because I don't see a sane way how this session can be assured to be present on boot) than the service should make its own login session. One way to do this would be a systemd user unit for root itself. The systemd user unit session could be started/stopped by this systemd service itself. It might be also possible to achieve the same using `systemd-run`.

Reply:
Hmm, indeed it may, even if this `ExecStartPre=` is here only for the `gpg`/`gpg-preset-passphrase` calls to upload material, not for the `gpg-agent` (using `systemd.sockets.gpg-agent`) itself, nor for the `secret-*.service` (which redo the same in their loop) if no `secret-*.service` remain in the `activating` state, the `/run/user/0` could disappear for good if root logins/logouts, at this state it's ok because no `/run/user/0` also redirects to the right socket but when login again to upload the OpenPGP material, there would be a `/run/user/0` but no `/run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts` and the connection to the right `gpg-agent ` would fail.
The obvious solution I see to this is to require to run `install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts` in the upload command before calling `gpg`/`gpg-preset-passphrase`, it would be equivalent to setting a custom `GPG_AGENT_INFO` on old `gpg`.
Concerning the "user-configured gpg-daemon" you would not recommend, well, if you don't recommend it I'll stay with requiring this `install` on upload :\

Reply:
I think your install command can fail in the following case:

1. No tmpfs exists at /run/user/0, your service runs: `install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts`
2. A root user logins in and `/run/user/0` is mounted

I will probably still works because gpg starts a new daemon but now you have two gpg-daemons.
So the question is why not just set TMPDIR to /tmp and use PrivateTmp=true? Than gpg should use this directory but the daemon is private to your service and its socket not accessible to the outside.

Reply:
Ok. That does not work either because `/run` will be always used.

```
$ gpgconf --list-dirs
sysconfdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/etc/gnupg
bindir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/bin
libexecdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/libexec
libdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/lib/gnupg
datadir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/share/gnupg
localedir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/share/locale
socketdir:/run/user/1000/gnupg
dirmngr-socket:/run/user/1000/gnupg/S.dirmngr
agent-ssh-socket:/run/user/1000/gnupg/S.gpg-agent.ssh
agent-extra-socket:/run/user/1000/gnupg/S.gpg-agent.extra
agent-browser-socket:/run/user/1000/gnupg/S.gpg-agent.browser
agent-socket:/run/user/1000/gnupg/S.gpg-agent
homedir:/home/joerg/.gnupg
```

I am now actually in favor of:

```nix
BindPaths = ["/$sometmpfs:/run/user/0"];
```

This seems a lot less hacky.

Reply:
Because of gpg-daemon I tried to avoid all gpg shenanigans in `sops-nix`.

Reply:
`TMPDIR` does not work because `/run/user` is hardcoded in gpg https://github.com/gpg/gnupg/blob/0904b8ef348a52335c378bee6dc90a978885d66f/common/homedir.c 
If the `install` is run just before the uploading `gpg` it's improbable that a root login could race it by mounting `/run/user/0`. 
Also,Maybe there is also a way to have this `install` run when root logins, by putting it in a `profile` or else.
Using `BindPaths=` would only work within the scope of the unit, and the uploading would have to use `unshare --mount sh -c 'mount --bind /$sometmpfs /run/user && gpg …'`, note that it would be `BindPaths = ["/$sometmpfs:/run/user"];` not `BindPaths = ["/$sometmpfs:/run/user/0"];` because `/run/user/0` must not exist for `gpg` to default to `${homedir}/S.gpg-agent`. Yes `gpg` is a pain on this point, or I'm trying to have a persistent gpg-daemon the wrong way and should indeed consider a user session, I don't know, anyway, I hope it goes smoother for you with `sops`.

Reply:
How about using `machinectl shell` to get a root logind session?

```
sudo machinectl shell .host $(which env) | grep RUNTIME
Connected to the local host. Press ^] three times within 1s to exit session.
XDG_RUNTIME_DIR=/run/user/0
```